### PR TITLE
Preserve empty NodeStateTable entries

### LINKED ISSUE #66324: ForAttributeWithMetadataName in IIncrementalGenerator never runs comparisons on output values
When using `ForAttributeWithMetadataName` rather than `CreateSyntaxProvider` the output values are never checked if equal, resulting in no caching and the function specified in `RegisterSourceOutput` being run for every change which matches the `predicate`. (Code Below)

**Version Used**: 
```xml
    <PropertyGroup>
        <TargetFramework>netstandard2.0</TargetFramework>
        <LangVersion>preview</LangVersion>
        <Nullable>enable</Nullable>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.4.0" PrivateAssets="all" />
        <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.3" PrivateAssets="All" />
    </ItemGroup
```
```txt
dotnet --info
.NET SDK:
 Version:   7.0.101
 Commit:    bb24aafa11

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  22.04
 OS Platform: Linux
 RID:         ubuntu.22.04-x64
 Base Path:   /usr/share/dotnet/sdk/7.0.101/

Host:
  Version:      7.0.1
  Architecture: x64
  Commit:       97203d38ba

.NET SDKs installed:
  6.0.404 [/usr/share/dotnet/sdk]
  7.0.101 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.12 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.1 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.12 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/usr/lib/dotnet/dotnet6-6.0.108]
```

**Example Code**

```cs
public class MyGenerator : IIncrementalGenerator {
    public virtual void Initialize( IncrementalGeneratorInitializationContext context ) {
        // CreateSyntaxProvider: this will hit the IntComparer
        IncrementalValuesProvider<int> createSyntaxProviderOutputValues =
                context.SyntaxProvider
                       .CreateSyntaxProvider( 
                           // simple filter for attributes
                           predicate: static ( node, _ ) => {
                               return node is ClassDeclarationSyntax { Identifier.ValueText: "TestDependentClass" };
                           },
                           // filter out attributed classes that we don't care about
                           transform: static ( syntaxContext, _ ) => {
                               return ( syntaxContext.Node as ClassDeclarationSyntax )?.Identifier.ValueText.Length ?? 33;
                           } )
                       .WithComparer( new IntComparer() );

        // this will NOT hit the IntComparer
        IncrementalValuesProvider<int> forAttributeWithMetadataNameOutputValues =
                context.SyntaxProvider
                       .ForAttributeWithMetadataName(
                           _attributeName,
                           // simple filter for attributes
                           predicate: static ( node, _ ) => {
                               return node is ClassDeclarationSyntax { Identifier.ValueText: "TestDependentClass" };
                           },
                           // filter out attributed classes that we don't care about
                           transform: static ( syntaxContext, _ ) => {
                               return ( syntaxContext.TargetNode as ClassDeclarationSyntax )?.Identifier.ValueText.Length ?? 33;
                           } ).WithComparer( new IntComparer() );

        context.RegisterSourceOutput( createSyntaxProviderOutputValues, ( SourceProductionContext sourceProductionContext, int value ) => {
            // ONLY EXECUTED ONCE
            // generate the source code and add it to the output
            string result = $"public class s_{value} {{ }}";
            sourceProductionContext.AddSource( $"CreateSyntaxProviderOutputValues_TEST_STRING_{( value )}.g.cs", SourceText.From( result, Encoding.UTF8 ) );
        } );

        context.RegisterSourceOutput( forAttributeWithMetadataNameOutputValues, ( SourceProductionContext sourceProductionContext, int value ) => {
            // EXECUTED TWICE
            // generate the source code and add it to the output
            string result = $"public class s_{value} {{ }}";
            sourceProductionContext.AddSource( $"ForAttributeWithMetadataNameOutputValues_{( value )}.g.cs", SourceText.From( result, Encoding.UTF8 ) );
        } );
}
internal class IntComparer : IEqualityComparer<int> {
    public bool Equals( int x, int y )=> x == y;
    public int GetHashCode( int obj ) => obj.GetHashCode();
}
```

Test code:
```cs

[ UsesVerify ] // Add hooks for Verify into XUnit
public class SyntaxTreeChangeTests {
    [ Fact ]
    public void NoRelevantChangeTest( ) {
        ( string original, string changed ) testClassSource = (
            /* language=cs */ """
            namespace NpgsqlEfSqlGenerator.Example {
                using System.ComponentModel.DataAnnotations;
                class TestClass {
                    public int Id { get;                                 set; }
                    public required string  Name                  { get; set; }
                    public          string? NullableReferenceType { get; set; }
                    public          int?    NullableValueType     { get; set; }
                }
            }
            """,
                              /* language=cs */ """
            namespace NpgsqlEfSqlGenerator.Example {
                using System.ComponentModel.DataAnnotations;
                class TestClass {
                    [ Key ]
                    public int Id { get;                                 set; }
                    public required string  Name                  { get; set; }
                    public          string? NullableReferenceType { get; set; }
                    public          int?    NullableValueType     { get; set; }
                }
            }
            """ );
        string testDependentClassSource =
            // original
            /* language=cs */
            """
            #nullable enable
            namespace NpgsqlEfSqlGenerator.Example {
                using System.ComponentModel.DataAnnotations;
                [ GenerateSql ]
                public class TestDependentClass {
                    [ Key ]
                    public int DependentId { get;                                 set; }
                    public required string  DependentName                  { get; set; }
                    public          string? DependentNullableReferenceType { get; set; }
                    public          int?    DependentNullableValueType     { get; set; }
                }
            }
            """;

        var compilation = TestHelper.CreateCompilation<NpgsqlEfSqlGenerator>(
            CSharpSyntaxTree.ParseText( testClassSource.original, path: "./TestClass.cs" ),
            CSharpSyntaxTree.ParseText( testDependentClassSource, path: "./TestDependentClass.cs" )
        );
        var generator = new NpgsqlEfSqlGenerator();

        GeneratorDriver driver = CSharpGeneratorDriver.Create( generator );

        driver = driver.RunGeneratorsAndUpdateCompilation( compilation, out var outputCompilation, out var diagnostics );
        // diagnostics.Should().BeEmpty();

        compilation = compilation.ReplaceSyntaxTree( compilation.SyntaxTrees[ 0 ], CSharpSyntaxTree.ParseText( testClassSource.changed, path: "./TestClass.cs" ) );
        driver = driver.RunGeneratorsAndUpdateCompilation( compilation, out outputCompilation, out diagnostics );
        diagnostics.Should().BeEmpty();
    }
}
```

**Diagnostic Id**:

If this is a report about a bug in an analyzer, please include the diagnostic if possible (e.g. `"IDE0030"`).

**Expected Behavior**:

For the `CreateSyntaxProvider` pipeline and `ForAttributeWithMetadataName` to both have their source output code run a single time when a change is not made to a matching item.

**Actual Behavior**:

For the `CreateSyntaxProvider` pipeline (see _ONLY EXECUTED ONCE_) is run once as expected as is `IntComparer.Equals`.
However for `ForAttributeWithMetadataName` (see _EXECUTED TWICE_) is run twice and `IntComparer.Equals` is never hit.

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes #66324.

### CODE REVIEW

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs [hunk 1]
```diff
@@ -63,7 +63,8 @@ public partial struct SyntaxValueProvider
         // changed. CreateSyntaxProvider will have to rerun all incremental nodes since it passes along the
         // SemanticModel, and that model is updated whenever any tree changes (since it is tied to the compilation).
         var syntaxTreesProvider = _context.CompilationProvider
-            .SelectMany((compilation, cancellationToken) => GetSourceGeneratorInfo(syntaxHelper, compilation, cancellationToken))
+            .SelectMany(static (compilation, _) => compilation.CommonSyntaxTrees)
+            .Select((tree, cancellationToken) => (Tree: tree, Info: tree.GetSourceGeneratorInfo(syntaxHelper, cancellationToken)))
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs:67]: Both steps here should have a tracking name. And the tests should validate the values in both steps.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs [hunk 2]
```diff
@@ -123,32 +124,6 @@ public partial struct SyntaxValueProvider
...
-        {
-            var info = tree.GetSourceGeneratorInfo(syntaxHelper, cancellationToken);
-            if ((info & SourceGeneratorSyntaxTreeInfo.ContainsGlobalAliasesOrAttributeList) != 0)
-                builder.Add((tree, info));
-        }
-
-        return builder.MoveToImmutable();
-    }
-
     private static ImmutableArray<SyntaxNode> GetMatchingNodes(
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs:127]: Where did this code go?

Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs:127]: If you mean the removed method, it was only called in `.SelectMany((compilation, cancellationToken) => GetSourceGeneratorInfo(syntaxHelper, compilation, cancellationToken))` above which I replaced by two steps with equivalent behavior. Instead of always getting only syntax trees with attributes or global usings, I use two steps - first get all syntax trees, then filter those with attributes or global usings. See point 2 in PR description on why I think this is better.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs (hunk 1 above)
Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs:67]: Good point, will add.

**Comment by sharwell** [comment]
> After this PR, the pipeline in the second generation starts with `[], [D]`, so later steps are cached effectively.

If you keep adding and removing items, does this list grow indefinitely?

**Comment by jjonescz** [comment]
> If you keep adding and removing items, does this list grow indefinitely?

Not really. If you start with list of syntax trees `[A], [B]`, then filter `A` out, you get `[], [B]`. But if you completely remove the first syntax tree, you still get just `[B]`.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 3]
```diff
@@ -90,6 +90,26 @@ private NodeStateTable(ImmutableArray<TableEntry> states, ImmutableArray<Increme
         public int GetTotalEntryItemCount()
             => _states.Sum(static e => e.Count);
 
+        public int GetTotalEntryItemCountPlusEmpty()
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:93]: The name doesn't quite seem to match up the logic here. The logic seems to be excluding trailing empty while the name implies it's including all empty. Suggest a name like `GetTotalEntryItemCountExceptTrailingEmpty` (might be a bit wordy though).

**Comment by CyrusNajmabadi** [comment]
@jjonescz what's the status of things here?  Thanks!

**Comment by jjonescz** [comment]
> what's the status of things here? Thanks!

@CyrusNajmabadi from what you wrote I got the feeling this is not a good optimization; wanted to eventually perf-test it as you suggested, but didn't get to that yet.
